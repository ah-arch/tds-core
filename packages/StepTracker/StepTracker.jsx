import React from 'react'
import PropTypes from 'prop-types'

import Text from '@tds/core-text'
import Box from '@tds/core-box'

import Step from './Step/Step'

import styles from './StepTracker.modules.scss'

/**
 * Show the current position in a sequence of steps.
 *
 * @version ./package.json
 */

const StepTracker = ({ current, steps, stepText }) => {
  return (
    <div>
      <Box vertical={2} dangerouslyAddClassName={styles.container}>
        {steps.map((label, index) => {
          return (
            /* eslint-disable react/no-array-index-key */
            <Step
              status={current}
              label={steps[index]}
              stepNumber={index + 1}
              stepIndex={index}
              key={index}
              data-testid="innerStep"
            />
            /* eslint-enable react/no-array-index-key */
          )
        })}
      </Box>
      <div className={styles.mobileLabel}>
        <Text>
          {stepText
            .replace('%current', current < steps.length ? current + 1 : steps.length)
            .replace('%total', steps.length)}{' '}
          {current < steps.length ? steps[current] : steps[steps.length - 1]}
        </Text>
      </div>
    </div>
  )
}

StepTracker.propTypes = {
  /**
   * The active step. The minimum value is 0, while the maximum value is the length of the steps prop.
   */

  // eslint-disable-next-line consistent-return
  current: PropTypes.number.isRequired,
  /**
   * The steps as an array of strings.
   */
  steps: PropTypes.array.isRequired,
  /**
   * String for text to display current step progress on mobile devices.
   * Use %current to place the current step and use %total to place the total amount of steps.
   */
  stepText: PropTypes.string,
}

StepTracker.defaultProps = {
  stepText: 'Step %current of %total:',
}

export default StepTracker
