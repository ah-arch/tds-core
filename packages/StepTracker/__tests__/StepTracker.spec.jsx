import React from 'react'
import { render, mount } from 'enzyme'

import Icon from '@tds/core-decorative-icon'

import StepTracker from '../StepTracker'

describe('<StepTracker />', () => {
  const defaultProps = {
    current: 1,
    steps: ['Login', 'Purchase', 'Checkout'],
  }

  const doMount = (overrides = {}) => {
    const container = mount(<StepTracker {...defaultProps} {...overrides} />)
    const innerSteps = container.find('[data-testid="innerStep"]')

    return {
      container,
      innerSteps,
    }
  }

  const doRender = (overrides = {}) => render(<StepTracker {...defaultProps} {...overrides} />)

  it('renders', () => {
    const stepTracker = doRender()

    expect(stepTracker).toMatchSnapshot()
  })

  it('contains defined steps', () => {
    const stepTracker = doMount()

    expect(stepTracker.innerSteps.length).toBe(3)
  })

  it('renders a single step in progress', () => {
    const { container, innerSteps } = doMount({ current: 2 })
    expect(container.find('div.stepActive').length).toBe(3)

    expect(innerSteps.at(0)).toContainReact(
      <Icon symbol="checkmark" size={16} variant="inverted" />
    )
    expect(innerSteps.at(1)).toContainReact(
      <Icon symbol="checkmark" size={16} variant="inverted" />
    )
    expect(innerSteps.at(2)).not.toContainReact(
      <Icon symbol="checkmark" size={16} variant="inverted" />
    )
  })
})
