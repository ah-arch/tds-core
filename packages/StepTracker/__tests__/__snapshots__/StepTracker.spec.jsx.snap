// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepTracker /> renders 1`] = `
<div>
  <div
    class="TDS_Box-modules__verticalPadding-2___1Uh2T container"
  >
    <div
      aria-current="false"
      aria-label="Login"
      class="stepActive"
    >
      <span
        class="icon"
      >
        <i
          aria-hidden="true"
          class="TDS_Icon-modules__iconCheckmark___3cAol TDS_Icon-modules__icon___13xYd TDS_Icon-modules__inverted___2mdlT TDS_Icon-modules__size16___1pm12"
        />
      </span>
      <span
        class="label"
      >
        <span
          class="TDS_Typography-modules__base___3JQ2A TDS_Typography-modules__wordBreak___11Xyp TDS_Typography-modules__baseFont___-Br-_ TDS_Typography-modules__color___22Swj"
        >
          1. Login
        </span>
      </span>
    </div>
    <div
      aria-current="true"
      aria-label="Purchase"
      class="stepActive"
    >
      <span
        class="icon"
      >
        <br />
      </span>
      <span
        class="label"
      >
        <span
          class="TDS_Typography-modules__base___3JQ2A TDS_Typography-modules__wordBreak___11Xyp TDS_Typography-modules__boldFont___30Cjc TDS_Typography-modules__wordBreak___11Xyp TDS_Typography-modules__color___22Swj"
        >
          2. Purchase
        </span>
      </span>
    </div>
    <div
      aria-current="false"
      aria-label="Checkout"
      class="step"
    >
      <span
        class="icon"
      >
        <br />
      </span>
      <span
        class="label"
      >
        <span
          class="TDS_Typography-modules__base___3JQ2A TDS_Typography-modules__wordBreak___11Xyp TDS_Typography-modules__baseFont___-Br-_ TDS_Typography-modules__color___22Swj"
        >
          3. Checkout
        </span>
      </span>
    </div>
  </div>
  <div
    class="mobileLabel"
  >
    <span
      class="TDS_Typography-modules__base___3JQ2A TDS_Typography-modules__wordBreak___11Xyp TDS_Typography-modules__baseFont___-Br-_ TDS_Typography-modules__color___22Swj"
    >
      Step 2 of 3: Purchase
    </span>
  </div>
</div>
`;
